pkg/builtin/unary/sin.go

```diff
@@ -0,0 +1,234 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package unary
+
+import (
+	"fmt"
+	"github.com/matrixorigin/matrixone/pkg/builtin"
+	"github.com/matrixorigin/matrixone/pkg/container/nulls"
+	"github.com/matrixorigin/matrixone/pkg/container/types"
+	"github.com/matrixorigin/matrixone/pkg/container/vector"
+	"github.com/matrixorigin/matrixone/pkg/encoding"
+	"github.com/matrixorigin/matrixone/pkg/sql/colexec/extend"
+	"github.com/matrixorigin/matrixone/pkg/sql/colexec/extend/overload"
+	"github.com/matrixorigin/matrixone/pkg/vectorize/sin"
+	"github.com/matrixorigin/matrixone/pkg/vm/process"
+)
+
+var argAndRets = []argsAndRet {
+	{[]types.T{types.T_uint8}, types.T_float64},
+	{[]types.T{types.T_uint16}, types.T_float64},
+	{[]types.T{types.T_uint32}, types.T_float64},
+	{[]types.T{types.T_uint64}, types.T_float64},
+
+	{[]types.T{types.T_int8}, types.T_float64},
+	{[]types.T{types.T_int16}, types.T_float64},
+	{[]types.T{types.T_int32}, types.T_float64},
+	{[]types.T{types.T_int64}, types.T_float64},
+
+	{[]types.T{types.T_float32}, types.T_float64},
+	{[]types.T{types.T_float64}, types.T_float64},
+}
+
+func init() {
+	extend.FunctionRegistry["sin"] = builtin.Sin	// register function name
+
+	for _, item := range argAndRets {
+		overload.AppendFunctionRets(builtin.Sin, item.args, item.ret)	// append function parameter types and return types
+	}
+	extend.UnaryReturnTypes[builtin.Sin] = func(extend extend.Extend) types.T {
+		return getUnaryReturnType(builtin.Sin, extend)	// define a get return type function for sin function
+	}
+
+	extend.UnaryStrings[builtin.Sin] = func(e extend.Extend) string {	// define a stringify function for sin
+		return fmt.Sprintf("sin(%s)", e)
+	}
+	overload.OpTypes[builtin.Sin] = overload.Unary	// register sin function type
+
+	overload.UnaryOps[builtin.Sin] = []*overload.UnaryOp {
+		{
+			Typ: types.T_float32,
+			ReturnType: types.T_float64,
+			Fn: func(origVec *vector.Vector, proc *process.Process, _ bool) (*vector.Vector, error) {
+				origVecCol := origVec.Col.([]float32)
+				resultVector, err := process.Get(proc, 8*int64(len(origVecCol)), types.Type{Oid: types.T_float64, Size: 8})
+				if err != nil {
+					return nil, err
+				}
+				results := encoding.DecodeFloat64Slice(resultVector.Data)
+				results = results[:len(origVecCol)]
+				resultVector.Col = results
+				nulls.Set(resultVector.Nsp, origVec.Nsp)
+				vector.SetCol(resultVector, sin.SinFloat32(origVecCol, results))
+				return resultVector, nil
+			},
+		},
+		{
+			Typ: types.T_float64,
+			ReturnType: types.T_float64,
+			Fn: func(origVec *vector.Vector, proc *process.Process, _ bool) (*vector.Vector, error) {
+				origVecCol := origVec.Col.([]float64)
+				resultVector, err := process.Get(proc, 8*int64(len(origVecCol)), types.Type{Oid: types.T_float64, Size: 8})
+				if err != nil {
```

Comment:
perhaps we could reuse the original vector when possible

Reply:
I have done this.